contracts/other/notional/Join1155.sol

```diff
@@ -0,0 +1,122 @@
+    function onERC1155Received(
+        address _operator,
+        address _from,
+        uint256 _id,
+        uint256 _value,
+        bytes calldata _data
+    ) external override returns(bytes4) {
+        require (_id == id, "Token id not accepted");
+        return ERC1155TokenReceiver.onERC1155Received.selector;
+    }
+
+    /// @dev Called by the sender after a batch transfer to verify it was received. Ensures only `id` tokens are received.
+    function onERC1155BatchReceived(
+        address _operator,
+        address _from,
+        uint256[] calldata _ids,
+        uint256[] calldata _values,
+        bytes calldata _data
+    ) external override returns(bytes4) {
+        uint256 length = _ids.length;
+        for (uint256 i; i < length; ++i)
+            require (_ids[i] == id, "Token id not accepted");
+        return ERC1155TokenReceiver.onERC1155BatchReceived.selector;
+    }
+
+    /// @dev Take `amount` `asset` from `user` using `transferFrom`, minus any unaccounted `asset` in this contract.
+    function join(address user, uint128 amount)
+        external override
+        auth
+        returns (uint128)
+    {
+        return _join(user, amount);
+    }
+
+    /// @dev Take `amount` `asset` from `user` using `transferFrom`, minus any unaccounted `asset` in this contract.
+    function _join(address user, uint128 amount)
+        internal
+        returns (uint128)
+    {
+        ERC1155 token = ERC1155(asset);
+        uint256 _storedBalance = storedBalance;
+        uint256 available = token.balanceOf(address(this), id) - _storedBalance; // Fine to panic if this underflows
+        storedBalance = _storedBalance + amount;
+        unchecked { if (available < amount) token.safeTransferFrom(user, address(this), id, amount - available, ""); }
+        return amount;        
+    }
+
+    /// @dev Transfer `amount` `asset` to `user`
+    function exit(address user, uint128 amount)
+        external override
+        auth
+        returns (uint128)
+    {
+        return _exit(user, amount);
+    }
+
+    /// @dev Transfer `amount` `asset` to `user`
+    function _exit(address user, uint128 amount)
+        internal
+        returns (uint128)
+    {
+        ERC1155 token = ERC1155(asset);
+        storedBalance -= amount;
+        token.safeTransferFrom(address(this), user, id, amount, "");
+        return amount;
+    }
+
+    /// @dev Retrieve any ERC1155 tokens other than the `asset`. Useful for airdropped tokens.
+    function retrieve(ERC1155 token, uint256 id_, address to)
+        external
+        auth
+    {
+        require(address(token) != address(asset) || id_ != id, "Use exit for asset");
+        token.safeTransferFrom(address(this), to, id_, token.balanceOf(address(this), id_), "");
+    }
+
+    /// @dev Retrieve any ERC20 tokens. Useful for airdropped tokens.
+    function retrieveERC20(IERC20 token, address to)
```

Comment:
one idea we could consider to make this a little more consistent with regular join is we could also name this fn `retrieve` as well and it would choose the fn based on the args passed in.  If we do that not sure if we should switch order or `id` and `to` -- another idea would be call this one `retrieve` and rename the other one `retrieve1155`.   but either way just a nit as these fns are not even in the interface

Reply:
I took the second option, as overloading is still quite clunky with ethers.